core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -0,0 +1,280 @@
+        props.setProperty(configName, Base64.encoder.encodeToString(value.getBytes(StandardCharsets.UTF_8)))
+    }
+    dynamicPasswordConfigs.foreach(convertToBase64)
+    props
+  }
+
+  private[server] def fromPersistentProps(persistentProps: Properties): Properties = {
+    val props = persistentProps.clone().asInstanceOf[Properties]
+    // TODO (KAFKA-6246): encrypt passwords
+    def convertFromBase64(configName: String): Unit = {
+      val value = props.getProperty(configName)
+      if (value != null)
+        props.setProperty(configName, new String(Base64.decoder.decode(value), StandardCharsets.UTF_8))
+    }
+    dynamicPasswordConfigs.foreach(convertFromBase64)
+    props
+  }
+}
+
+class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging {
+
+  private[server] val staticBrokerConfigs = configToMap(kafkaConfig.originalsFromThisConfig.asScala)
+  private[server] val staticDefaultConfigs = configToMap(KafkaConfig.defaultValues)
+  private[server] val dynamicBrokerConfigs = mutable.Map[String, String]()
+  private[server] val dynamicDefaultConfigs = mutable.Map[String, String]()
+  private val brokerId = kafkaConfig.brokerId
+  private val reconfigurables = mutable.Buffer[Reconfigurable]()
+  private val lock = new ReentrantReadWriteLock
+  private var currentConfig = kafkaConfig
+
+  def initialize(zkClient: KafkaZkClient): Unit = {
+    val adminZkClient = new AdminZkClient(zkClient)
+    updateDefaultConfig(adminZkClient.fetchEntityConfig(ConfigType.Broker, ConfigEntityName.Default))
+    updateBrokerConfig(brokerId, adminZkClient.fetchEntityConfig(ConfigType.Broker, brokerId.toString))
+    updateCurrentConfig()
+  }
+
+  def config: KafkaConfig = CoreUtils.inReadLock(lock) {
+    currentConfig
+  }
+
+  def addReconfigurable(reconfigurable: Reconfigurable): Unit = CoreUtils.inWriteLock(lock) {
+    require(reconfigurable.reconfigurableConfigs.asScala.forall(allDynamicConfigs.contains))
+    reconfigurables += reconfigurable
+  }
+
+  def removeReconfigurable(reconfigurable: Reconfigurable): Unit = CoreUtils.inWriteLock(lock) {
+    reconfigurables -= reconfigurable
+  }
+
+  private[server] def updateBrokerConfig(brokerId: Int, persistentProps: Properties): Unit = CoreUtils.inWriteLock(lock) {
+    val props = fromPersistentProps(persistentProps)
+    validateConfigTypes(Some(brokerId), props, logError = true)
+    dynamicBrokerConfigs.clear()
+    dynamicBrokerConfigs ++= props.asScala
+    updateCurrentConfig()
+  }
+
+  private[server] def updateDefaultConfig(persistentProps: Properties): Unit = CoreUtils.inWriteLock(lock) {
+    val props = fromPersistentProps(persistentProps)
+    validateConfigTypes(None, props, logError = true)
+    dynamicDefaultConfigs.clear()
+    dynamicDefaultConfigs ++= props.asScala
+    updateCurrentConfig()
+  }
+
+  private[server] def validateAndConvertForPersistence(brokerId: Option[Int], props: Properties): Properties = CoreUtils.inReadLock(lock) {
```

Comment:
nit: since there is only one caller, it seems like we could just separate validation and conversion.